### PR TITLE
Strip double # char on HexInput

### LINKED ISSUE #38334: Pasting hex value into color picker defaults to black
### Description

When pasting a hex value containing the `#`, the character isn't removed and a black color is computed.

### Step-by-step reproduction instructions

1. Add a block
2. Choose a color
3. Edit the hex value
4. Paste in a value with `#` at the start

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/1478421/151591457-f6117446-0dd5-437b-82a1-7e2ce30ad5a7.mov

### Environment info

- WP 5.9
- Gutenberg 12.5.0-rc.1 
- TT1 Blocks theme

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by KevinBatdorf**
## Description
This strips out the leading `#` when setting a color hex value on the editor

Closes https://github.com/WordPress/gutenberg/issues/38334

## Testing Instructions

1. Add a block
2. Choose a color
3. Edit the hex value
4. Paste in a value with # at the start

## Screenshots 

Check the issue linked above for a screencast of the previous behavior

https://user-images.githubusercontent.com/1478421/151591738-e2906fa8-bead-4163-8061-9fd6e2e1c721.mov

## Types of changes

Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [ ] I've updated related schemas if appropriate.

### CODE REVIEW

File: packages/components/src/color-picker/hex-input.tsx [hunk 1]
```diff
@@ -24,6 +24,10 @@ interface HexInputProps {
 }
 
 export const HexInput = ( { color, onChange, enableAlpha }: HexInputProps ) => {
+	const handleChange = ( nextValue ) => {
+		const hexValue = ( '#' + nextValue ).replace( '##', '#' );
+		onChange( colord( hexValue ) );
+	};
 	const handleValidate = ( value: string ) => {
```
Comment (gwwar) [inline packages/components/src/color-picker/hex-input.tsx]: I noticed this while reviewing that we never call the validation function. ColorHexInputControl needs the `isPressEnterToChange` property to check values.

Here's what the behavior looks like when this is set. If invalid values are entered, we keep the last color value. The tradeoff though is that the color value needs an <kbd>enter</kbd> keypress, versus updating as we type.

https://user-images.githubusercontent.com/1270189/153684366-5a83252c-4f6c-4d71-adb3-73abf7e4263a.mp4

File: packages/components/src/color-picker/hex-input.tsx [hunk 2]
```diff
@@ -24,6 +24,10 @@ interface HexInputProps {
 }
 
 export const HexInput = ( { color, onChange, enableAlpha }: HexInputProps ) => {
+	const handleChange = ( nextValue ) => {
+		const hexValue = ( '#' + nextValue ).replace( '##', '#' );
```
Comment (gwwar) [inline packages/components/src/color-picker/hex-input.tsx]: This is more minor, but I'd recommend checking if `#` already is in nextValue, or stripping all `#` before adding a prefix.

**Review by gwwar — COMMENTED** [review]
Thanks for your contribution @KevinBatdorf ! I think this is a pretty reasonable change to make. 

Let's check in with @sarayourfriend and @jorgefilipecosta on what the intended validation behavior was meant to be for this field. I noticed while reviewing that we never call the validation function. ColorHexInputControl needs the `isPressEnterToChange` property to check values.

File: packages/components/src/color-picker/hex-input.tsx (hunk 2 above)
Comment (KevinBatdorf) [reply packages/components/src/color-picker/hex-input.tsx]: I switched it to use a ternary and handle undefined.

File: packages/components/src/color-picker/hex-input.tsx (hunk 1 above)
Comment (KevinBatdorf) [reply packages/components/src/color-picker/hex-input.tsx]: Are you able to actually trigger the validation handler? Even with bad input I can't get it to throw the error. The console remains empty.

I updated it to include `isPressEnterToChange` and also added an `onPaste` handler that will trigger the `handleChange` function. 

What do you think about that? 

If `onValidate` isn't working as expected, should we just throw on bad input within `handleChange`?

Comment (gwwar) [reply packages/components/src/color-picker/hex-input.tsx]: Oh so I don't think we need to change the behavior to add `isPressEnterToChange`, I was just noting that the `handleValidate` doesn't do anything. Unless other folks chime in, I think it's okay to remove the `handleValidate` handler, so folks don't wonder why nothing is happening. https://github.com/WordPress/gutenberg/blob/ea32685fbabf070e022112b0b94c9e4cfa8ff307/packages/components/src/input-control/input-field.tsx#L160-L162

> Are you able to actually trigger the validation handler? 

Yes, I'm able to. One thing I am running into is that the max length set on the input will truncate the values we are interested in. This appears to be the case for osx / Chrome,FF,Safari combos. Which os/browser are you using? I wonder if we're seeing some subtle browser differences.

https://user-images.githubusercontent.com/1270189/153907935-9427ed6e-ff73-4234-add2-2f78ec1f81e2.mp4

**Comment by gwwar** [comment]
I think we're pretty close. Playing around with this, it gets a little tricky if we start incorporating the onPaste handler. If we're trying to modify paste events, we normally preventDefault, then manually update the input.

What about the following?
- Try changing maxLength to 7 for hex and accept both inputs like `#55c3dc` and `55c3dc`
- Keep changes to `handleChange`
- Remove `handleValidate` (since it only works with `isPressEnterToChange` set)
- Drop onPaste handler
- Remove `isPressEnterToChange` prop to retain current behavior.

It should look like this all together:

https://user-images.githubusercontent.com/1270189/153916765-dea1ffa1-f2fb-4d9e-b0f3-b008b8e9ba05.mp4

The only slightly finicky thing is that we show another `#`, but that only remains until we clear focus.

cc @jasmussen for the proposed behavior changes, of max input length from 6 to 7

File: packages/components/src/color-picker/hex-input.tsx (hunk 2 above)
Comment (gwwar) [reply packages/components/src/color-picker/hex-input.tsx]: Thanks that looks good to me!

**Comment by jasmussen** [comment]
> cc @jasmussen for the proposed behavior changes, of max input length from 6 to 7

Appreciate the ping, thank you. Judging only by the video, this seems like a totally fine sanitization to me. In fact I often find myself frustrated with inputs that truncate hex codes due to the addition of the hash mark ⭐

**Comment by jorgefilipecosta** [comment]
> Let's check in with @sarayourfriend and @jorgefilipecosta on what the intended validation behavior was meant to be for this field. 

I'm not inside of the intended validation behavior but the plan @gwwar suggested seems good to me (provided in max length we handle the case where alpha is enabled).

**Comment by gwwar** [comment]
Excellent, sounds like we have a plan. Once we have changes on the branch, it should be pretty quick to retest, approve, and merge. @KevinBatdorf let me know if anything is unclear, or if you'd prefer for me to add changes directly to the branch.

**Comment by KevinBatdorf** [comment]
Hi everyone. Thanks for the feedback. I'll most likely follow up on this tomorrow or if not tomorrow then definitely Friday (we are doing our FFTF this Friday!).

**Comment by KevinBatdorf** [comment]
Hi @gwwar I added these changes:

```
✅ Try changing maxLength to 7 for hex and accept both inputs like #55c3dc and 55c3dc
✅ Keep changes to handleChange
✅ Remove handleValidate (since it only works with isPressEnterToChange set)
✅ Drop onPaste handler
✅ Remove isPressEnterToChange prop to retain current behavior.
```

I also updated the max length for when `enableAlpha` is set to 9

>  if anything is unclear, or if you'd prefer for me to add changes directly to the branch.

Feel free. I read the messages in a rush the other day so missed this part, but overall it's always okay with me if someone wants to take over a branch here. I'm trying to jump in during my free time so I don't _yet_ have a full understanding of all the components and how they interact with each other. If you want to make some more changes feel free! 

Thanks for the feedback and everything else as well.

**Review by gwwar — APPROVED** [review]
Thanks @KevinBatdorf for the contribution! This tests very well for me. Here's what I see:

https://user-images.githubusercontent.com/1270189/154722416-62566c02-0683-424d-9ad4-fd07b0de6961.mp4

I'll merge this for you once CI tests are ✅

**Comment by KevinBatdorf** [comment]
Thanks @gwwar One optimization I think would make it even better is if it would auto focus the input when you press the control button. I may explore that later on.

**Pull request merged**